python/preprocessor.py

```diff
@@ -0,0 +1,349 @@
+
+        returns
+        -------
+        an 1d numpy array of boolean indicating whether each index is to be modified
+        """
+        mask = np.zeros(np.shape(observables))
+        if all_observable_key in modify:
+            # indicating all observables should be preprocessed
+            mask = mask + 1 # set every mask item to 1
+        else:
+            for ob_name in modify:
+                mask[observables == ob_name] = 1
+        mask = mask == 1
+        return mask
+
+    def preprocess(self, features, feature_array, **args):
+        """
+        preprocess feature_array by applying requested preprocessor functions
+
+        arguments
+        ---------
+        features: name of root tree branchs
+        feature_array: 2d numpy array with shape (number of events, observables in each event)
+        args: extra parameters that will be passed directly to supported preprocessor functions
+
+        returns
+        -------
+        preprocessed feature array
+        """
+        # convert feature names to observable names
+        observables = np.array([self.observable_name_dict[feature] for feature in features])
+
+        # use as a checklist to mark the items that are done
+        task_list = (self.config[FEATURE]).copy()
+
+        logger.debug("Observable order before preprocessing: "+str(observables))
+
+        while task_list:
+            function_name, list_of_observable = task_list.popitem(last = False)
+            logger.debug("Applying preprocessing function " + function_name)
+            function = self.feature_preprocessing_function_map[function_name]
+
+            # create a mask for the next operation
+            mask = self._mask(observables, list_of_observable)
+
+            # call preprocessor function, passing any additional args as is
+            # modify args here to add in additional arguments passed to preprocessor function
+            feature_array, observables = function(feature_array, mask, observables, **args)
+
+            logger.debug("Observable order after preprocessing round: "+str(observables))
+
+        gc.collect()
+        
+        # return the feature array after preprocessing
+        return feature_array
+    
+    def preprocess_weight(self, feature_arrays, weights, observables, **args):
+        """
+        apply the list of weight preprocessors in sequence and return the final set of preprocessed weights
+
+        arguments
+        ---------
+        feature_arrays: tuple of numpy arrays
+            feature arrays in the order of [data, sim, gen]
+        weights: tuple of numpy arrays
+            event weights in the same order as feature arrays
+        observables: list of str
+            indicating the position of each observable in feature array
+        """
+
+        for function in self.weight_preprocessing_functions:
+            weights = function(feature_arrays, weights, observables, **args)
+
+        return weights
+
+    def use_utility(self):
```

Comment:
So to replace the boolean with a str of the name? The idea I had in mind was that there might be several of these utilities in use, let's say preprocessing function A, B, and C. The boolean method would be saying do use utilities A, B, and C. It can get confusing forming a str to specifying using "A","B","C", unless we replace it with an array of str.

Then the fundamental problem is that these are functions that requires the user to decide things like which two arrays will be normalized together. They would still need to manually put in the code in omnifoldTTbar. Therefore, just specifying the names of the functions to preprocessor doesn't work.

I thought of passing all feature arrays to the preprocessor. However, then we need to specify the name of preprocessor and then how the arrays are paired for each of them. It is doable but I feel that it will quickly get confusing and is probably isn't much easier to write a function call.

Reply:
I was mainly thinking in case we'd like to switch between different methods e.g. divide by the order of magnitude, or divide by its max absolute value or divide by its std, etc., one then has to modify what `use_utility` does in the code.

As for which arrays are normalized together, this kind of preprocessing will always be applied to data and reco array togther (as X for step 1), and separately for array gen (as X for step2).

Reply:
I can do a midway approach for this. We can request functions as a list of str. The default pairing would be data + reco and gen alone. It should satisfy practically all situations. If other kind of pairing is needed, then directly editing the code will be needed. What do you think of this approach?

Reply:
Sounds reasonable to me! 
I'm not really sure about this but another idea is to apply this directly on `X_step1` and `X_step2` (`X_step1b`, `X_step2b` too I guess) in `omnifold.py`. If we do this we also don't need to worry about the `group_standardize` above.

Reply:
Cancel that and see the next thread. I think moving 2nd preprocessing step into omnifold solves this problem.